java/src/main/java/com/microsoft/onnxruntime/ONNXAllocator.java

```diff
@@ -0,0 +1,294 @@
+     *
+     * If the supplied buffer is not a direct buffer, a direct copy is created tied to the lifetime
+     * of the tensor.
+     * @param data The tensor data.
+     * @param shape The shape of tensor.
+     * @return An ONNXTensor of the required shape.
+     * @throws ONNXException Thrown if there is an onnx error or if the data and shape don't match.
+     */
+    public ONNXTensor createTensor(IntBuffer data, long[] shape) throws ONNXException {
+        if (!closed) {
+            ONNXJavaType type = ONNXJavaType.INT32;
+            int bufferSize = data.capacity()*type.size;
+            IntBuffer tmp;
+            if (data.isDirect()) {
+                tmp = data;
+            } else {
+                ByteBuffer buffer = ByteBuffer.allocateDirect(bufferSize).order(ByteOrder.nativeOrder());
+                tmp = buffer.asIntBuffer();
+                tmp.put(data);
+            }
+            TensorInfo info = TensorInfo.constructFromBuffer(tmp, shape, type);
+            return new ONNXTensor(createTensorFromBuffer(ONNX.ortApiHandle, handle, tmp, bufferSize, shape, info.onnxType.value), handle, info, tmp);
+        } else {
+            throw new IllegalStateException("Trying to create an ONNXTensor on a closed ONNXSession.");
+        }
+    }
+
+    /**
+     * Create an ONNXTensor backed by a direct LongBuffer.
+     *
+     * If the supplied buffer is not a direct buffer, a direct copy is created tied to the lifetime
+     * of the tensor.
+     * @param data The tensor data.
+     * @param shape The shape of tensor.
+     * @return An ONNXTensor of the required shape.
+     * @throws ONNXException Thrown if there is an onnx error or if the data and shape don't match.
+     */
+    public ONNXTensor createTensor(LongBuffer data, long[] shape) throws ONNXException {
+        if (!closed) {
+            ONNXJavaType type = ONNXJavaType.INT64;
+            int bufferSize = data.capacity()*type.size;
+            LongBuffer tmp;
+            if (data.isDirect()) {
+                tmp = data;
+            } else {
+                ByteBuffer buffer = ByteBuffer.allocateDirect(bufferSize).order(ByteOrder.nativeOrder());
+                tmp = buffer.asLongBuffer();
+                tmp.put(data);
+            }
+            TensorInfo info = TensorInfo.constructFromBuffer(tmp, shape, type);
+            return new ONNXTensor(createTensorFromBuffer(ONNX.ortApiHandle, handle, tmp, bufferSize, shape, info.onnxType.value), handle, info, tmp);
+        } else {
+            throw new IllegalStateException("Trying to create an ONNXTensor on a closed ONNXSession.");
+        }
+    }
+
+    /**
+     * Closes the allocator, must be done after all it's child objects have been closed.
+     * @throws ONNXException If it failed to close.
+     */
+    @Override
+    public void close() throws ONNXException {
+        if (!closed) {
+            //closeAllocator(ONNX.ortApiHandle,handle);
+            closed = true;
```

Comment:
>closed = true [](start = 12, length = 13)

Native handle must be closed

Reply:
The free method for an allocator disappeared between 0.5 and the current master. I'm not sure what C API endpoint I should use to free one.

Reply:
Will  follow up

---
In reply to: [340852616](https://github.com/microsoft/onnxruntime/pull/2215#discussion_r340852616) [](ancestors = 340852616)

Reply:
This class houses a Default allocator which is a singlton. No need to free it. However, I suggest you reflect in the class name and in the comments.
You may want to create a separate ONNXAllocator class that creates and frees custom alloctors. For that you can allocate OrtAlloctor structure, fill in members and use it. You'd have to free it when it is done.

---
In reply to: [340875164](https://github.com/microsoft/onnxruntime/pull/2215#discussion_r340875164) [](ancestors = 340875164,340852616)

Reply:
Ok, I can do that. Is there an example somewhere of making a non default allocator? All that stuff changed recently and I'm not sure how best to use the CAPI.

Reply:
I can't find it handy. We can add custom allocator later.

---
In reply to: [340885378](https://github.com/microsoft/onnxruntime/pull/2215#discussion_r340885378) [](ancestors = 340885378)